### PR TITLE
Add stale branches heads to finality notifications

### LINKED ISSUE #10605: Extend finality notifications to include displaced leaves
The client provides a stream for listening for finality notifications, thus allowing other clients to hook into the finality pipeline. Currently it sends a notification for each block that gets finalized with the following format:

https://github.com/paritytech/substrate/blob/7409589ab78e7cb508db873dd6384634d667b8f5/client/api/src/client.rs#L273-L280

Finalizing a block can lead to a bunch of forks becoming stale (i.e. they cannot progress any further). In such cases, it might be useful to know that these forks can be abandoned so that potentially some data can be pruned (e.g. BABE block weight). Currently, this is handled in the backend where we keep track of the open leaves but this information is not exposed back to the client. I propose updating the finality notifcation struct to include this data:

```rust
pub struct FinalityNotification<Block: BlockT> {
    pub finalized_hash: Block::Hash,
    pub finalized_header: Block::Header,
    pub previous_finalized_number: NumberFor<Block>,
    pub stale_heads: Vec<Block::Hash>,
}
```

`stale_heads` would include the latest block of each of these forks, all stale fork blocks could then be found by traversing these heads back until reaching the finalized chain.

Additionally, I suggest that the notification should also include the block number of the previously best finalized block. This is because currently the client will create notifications for all intermediary finalized blocks, e.g. if block 5 is the current best finalized and we then finalize block 10 then the client will send notifications for 6, 7, 8, 9 and 10. By passing along the last finalized number in the notification we can avoid having to create all these intermediary notifications and instead the consumer of the notification can decide whether it needs to do anything on the intermediary blocks or not. The current implementation also has a edge case where we don't send more than 256 notifications, which can lead to some unexpected results (https://github.com/paritytech/substrate/blob/master/client/service/src/client/client.rs#L833-L842).

### PR DESCRIPTION
**Pull request by davxy**
Closes #10605 

Includes stale branches heads to finality notifications sent by the client.

**Warning**

Previous implementation was sending a notification for each block between the previous (explicitly) finalized block and
the new finalized one (with an hardcoded limit of 256). Thus finalization messages were sent also for implicitly finalized blocks.

Now finality notification is sent only for the new finalized head and it contains the hash of the new finalized head hash, new finalized head header, a list of all the implicitly finalized blocks and a list of stale branches heads (i.e. heads that are not part of the canonical chain anymore).

### CODE REVIEW

File: client/service/src/client/client.rs [hunk 1]
```diff
@@ -831,58 +843,59 @@ where
...
-		for finalized_hash in notify_finalized {
-			let header = self.header(&BlockId::Hash(finalized_hash))?.expect(
-				"Header already known to exist in DB because it is indicated in the tree route; \
-				 qed",
-			);
+		let mut stale_heads = self.backend.blockchain().leaves()?;
+		stale_heads.retain(|block| {
+			let route_from_finalized =
+				sp_blockchain::tree_route(self.backend.blockchain(), notify_finalized.hash, *block)
+					.unwrap();
```
Comment (bkchr) [inline client/service/src/client/client.rs:887]: No unwrap. Either `expect` with a proof that this can not fail or handle it gracefully.

File: client/api/src/client.rs [hunk 2]
```diff
@@ -273,10 +273,14 @@ pub struct BlockImportNotification<Block: BlockT> {
...
 #[derive(Clone, Debug)]
 pub struct FinalityNotification<Block: BlockT> {
-	/// Imported block header hash.
+	/// Finalized block header hash.
 	pub hash: Block::Hash,
-	/// Imported block header.
+	/// Finalized block header.
 	pub header: Block::Header,
+	/// Previous explicitly finalized block header hash.
+	pub previous_finalized_hash: Block::Hash,
```
Comment (bkchr) [inline client/api/src/client.rs:281]: Maybe we should instead have a `Arc<[Block::Hash]>` here? Where the first is the previous last finalized and the last is the currently finalized one.

File: client/transaction-pool/api/src/lib.rs [hunk 3]
```diff
@@ -289,8 +289,10 @@ pub enum ChainEvent<B: BlockT> {
 	},
 	/// An existing block has been finalized.
 	Finalized {
-		/// Hash of just finalized block
+		/// Hash of just finalized block.
 		hash: B::Hash,
+		/// Path from old finalized to new finalized parent.
+		tree_route: Arc<Vec<B::Hash>>,
```
Comment (bkchr) [inline client/transaction-pool/api/src/lib.rs:295]:
```suggestion
		/// The hashes of all finalized blocks.
		finalized: Arc<[B::Hash]>,
```

File: client/api/src/client.rs [hunk 4]
```diff
@@ -273,10 +273,14 @@ pub struct BlockImportNotification<Block: BlockT> {
...
-	/// Imported block header hash.
+	/// Finalized block header hash.
 	pub hash: Block::Hash,
-	/// Imported block header.
+	/// Finalized block header.
 	pub header: Block::Header,
+	/// Path from the old finalized to new finalized parent (implicitly finalized blocks).
+	pub tree_route: Arc<Vec<Block::Hash>>,
+	/// Stale branches heads.
+	pub stale_heads: Arc<Vec<Block::Hash>>,
```
Comment (bkchr) [inline client/api/src/client.rs:283]:
```suggestion
	pub tree_route: Arc<[B::Hash]>,
	/// Stale branches heads.
	pub stale_heads: Arc<[B::Hash]>,
```

**Comment by davxy** [comment]
@bkchr @andresilva 

### TL;DR

A fairly meticulous inspection of `finality_notification_stream` consumers; 

In particular assumptions about sequential block finalization messages were checked and eventually fixed.

 **Looks like** receiving the last finalized block message was sufficient for almost all consumers with the exception of txpool, where a fix is already contained in this PR. 
 
A review from more experienced people for the single components is still required.

Follows a recap for Substrate, Cumulus and Polkadot with code references to simplify the inspection.


# Substrate

### Transaction Pool

- Stream read by `notification_future` [here](https://github.com/paritytech/substrate/blob/55e470a0c44ec35d37f0221fc44e4189c0037b24/client/transaction-pool/src/lib.rs#L732-L747)
- Validated transaction pool assumes to receive a notification message for each finalized block.
- Message handler has been adjusted to send the full finalized list to `pool.validated_pool().on_block_finalized()`

See [here](https://github.com/paritytech/substrate/blob/55e470a0c44ec35d37f0221fc44e4189c0037b24/client/transaction-pool/src/lib.rs#L715-L723)

### Beefy

- Stream read by `BeefyWorker` [here](https://github.com/paritytech/substrate/blob/55e470a0c44ec35d37f0221fc44e4189c0037b24/client/beefy/src/worker.rs#L407-L412)
- Message is then processed [here](https://github.com/paritytech/substrate/blob/55e470a0c44ec35d37f0221fc44e4189c0037b24/client/beefy/src/worker.rs#L215)

- Looks happy with final head notification.

### Networking

- Stream is read within `build_network_future` [here](https://github.com/paritytech/substrate/blob/55e470a0c44ec35d37f0221fc44e4189c0037b24/client/service/src/lib.rs#L155-L156)
- `build_network_future` forwards finality notification to `network.on_block_finalized(hash, number)` [here](https://github.com/paritytech/substrate/blob/55e470a0c44ec35d37f0221fc44e4189c0037b24/client/service/src/lib.rs#L200-L201)
- message arrives to `ChainSync::on_block_finalized` [here](https://github.com/paritytech/substrate/blob/55e470a0c44ec35d37f0221fc44e4189c0037b24/client/network/src/protocol/sync.rs#L1646).
- message is forwarded to `ExtraRequests::on_block_finalized` in [here](https://github.com/paritytech/substrate/blob/55e470a0c44ec35d37f0221fc44e4189c0037b24/client/network/src/protocol/sync/extra_requests.rs#L160)

- **Seems happy with final head notification but a meticuolous review required**.

### RPC

- Stream read messages are forwarded to subscribers of `subscribe_finalized_heads` API [here](https://github.com/paritytech/substrate/blob/55e470a0c44ec35d37f0221fc44e4189c0037b24/client/rpc/src/chain/mod.rs#L162-L178)

- **WARNING**: we are now sending notifications only for finalized heads but external applications may rely on previous behaviour?


# Cumulus

### PovRecovery

- Stream is read within `PovRecovery::run` [here](https://github.com/paritytech/cumulus/blob/3994f6cbeca18ac37c9b6b1d04c5b47004eb2687/client/pov-recovery/src/lib.rs#L393-L395)
- Calls `handle_block_finalized(finalized_number)` that is used to remove from `self.pending_candidates` blocks with `number <= finalized_number` [here](https://github.com/paritytech/cumulus/blob/3994f6cbeca18ac37c9b6b1d04c5b47004eb2687/client/pov-recovery/src/lib.rs#L207-L209)

- **Seems happy with final head notification**

### RelaychainClient

- Stream is read within `RelaychainClient for RCInterface :: finalized_heads` implementation [here](https://github.com/paritytech/cumulus/blob/3994f6cbeca18ac37c9b6b1d04c5b47004eb2687/client/consensus/common/src/parachain_consensus.rs#L393-L403)
- The stream result is then used in `follow_finalized_head`. [here](https://github.com/paritytech/cumulus/blob/3994f6cbeca18ac37c9b6b1d04c5b47004eb2687/client/consensus/common/src/parachain_consensus.rs#L72)
    - for each new finalized head the  `parachain.finalize_block(finalized_hash)` is called.
    - this is already correctly handled by Substrate `Client` implementation for `Finalizer` trait.

- **Seems happy with final head notification**

# Polkadot

### Overseer

node/overseer/src/lib.rs
- `forward_events()` forwards stream messages to the overseer wrapper using `Handle::block_finalized` [here](https://github.com/paritytech/polkadot/blob/dee0681cccc650d7b2ba5face7f15eb49bdc9b98/node/overseer/src/lib.rs#L281)
- `Handle::block_finalized` sends messages to Overseer task.
- The event is catched within the `Overseer::run` and managed using `Overseer::block_finalized` [here](https://github.com/paritytech/polkadot/blob/dee0681cccc650d7b2ba5face7f15eb49bdc9b98/node/overseer/src/lib.rs#L621-L623)
   - This prunes leaves below the finalized head [here](https://github.com/paritytech/polkadot/blob/dee0681cccc650d7b2ba5face7f15eb49bdc9b98/node/overseer/src/lib.rs#L685)

**Comment by bkchr** [comment]
@davxy ty for the detailed report. Looks good as far as I can see :)

File: client/transaction-pool/api/src/lib.rs (hunk 3 above)
Comment (bkchr) [inline client/transaction-pool/api/src/lib.rs:295]: We should merge both imo. There is no real need to just have the latest finalized block.

File: client/service/src/lib.rs [hunk 5]
```diff
@@ -152,26 +152,7 @@ async fn build_network_future<
...
-			}
-			if let Some(last) = last {
-				Poll::Ready(Some(last))
-			} else {
-				Poll::Pending
-			}
-		})
-		.fuse()
-	};
+	let mut finality_notification_stream = client.finality_notification_stream().fuse();
```
Comment (andresilva) [inline client/service/src/lib.rs:155]: We should keep this behavior in this PR, but this could be a source of bugs at least for justification requests (I actually wonder if it can explain some rare bugs that have been reported). We should probably create an issue to adapt the `on_block_finalized` hooks, to take in the whole notification. This way, e.g. inside sync we could make sure that `extra_justifications.on_block_finalized` gets called for all finalized blocks.

File: client/service/src/client/client.rs [hunk 6]
```diff
@@ -622,25 +622,25 @@ where
...
 
 				operation.op.update_cache(new_cache);
 				storage_changes
 			},
 			None => None,
 		};
 
+		// Ensure parent chain is finalized to maintain invariant that finality is called
+		// sequentially.
+		if finalized && parent_exists {
```
Comment (andresilva) [inline client/service/src/client/client.rs:634]: Yeah this was a bug, there's no reason not to do this when there's no storage changes.

File: client/service/src/client/client.rs [hunk 7]
```diff
@@ -831,58 +847,73 @@ where
...
+				.blockchain()
+				.number(last_finalized)?
+				.expect("Finalized block expected to be onchain. qed");
+			let mut stale_heads = Vec::new();
+			for head in self.backend.blockchain().leaves()? {
+				let route_from_finalized =
+					sp_blockchain::tree_route(self.backend.blockchain(), block, head)?;
+				let retracted = route_from_finalized.retracted();
+				let pivot = route_from_finalized.common_block();
+				if !(retracted.is_empty() || last_finalized_number > pivot.number) {
```
Comment (andresilva) [inline client/service/src/client/client.rs:865]: Can you explain the second condition? If nothing is retracted then the common block should be `block`, in which case the `last_finalized_number` is never higher. Is my thinking correct here?

**Review by andresilva — COMMENTED** [review]
Good work @davxy, LGTM overall. For the RPC I'm not sure whether we should change it to take the summary and emit a notification for each block. We need to add some tests for the creation of stale heads because I am not sure if all cases are covered or whether we might be missing some heads.

@jacogr When multiple blocks are finalized at once, would getting a single finality notification for the latest block, as opposed to now getting one for each finalized block, have any effect on polkadot.js/apps?

File: client/service/src/client/client.rs (hunk 7 above)
Comment (davxy) [reply client/service/src/client/client.rs:865]: I spotted that `backend.blockchain().leaves()` may return heads that were already staled in previous finalization/s and thus already included in prev notifications.

Motivation: the `Backend` implementation for `BlockchainDb` stores the leaves in memory [here](https://github.com/paritytech/substrate/blob/4cd2cc958f5a078d87e4eae877167c40cf2ebf9e/client/db/src/lib.rs#L449-L456) and being stale is not a sufficient condition to immediately remove the head from the list, see [here](https://github.com/paritytech/substrate/blob/4cd2cc958f5a078d87e4eae877167c40cf2ebf9e/client/api/src/leaves.rs#L131-L134) why.

Said that, we want to avoid returning stale heads that were already included in previous finalizations.

Lets split the condition in two for easier analysis.
    `!retracted.is_empty() && !(last_finalized_number > pivot.number)`

To be added to stale heads list, first of all `retracted` from current finalized block to the head should not be empty.

But this last condition is not sufficient to avoid including heads that were already included in previous finalization calls.

An efficient way to check this condition is by controlling that the block height of the previously finalized block is not greater than the pivot (between new finalized and head we are checking), in other words:
- if last_finalized_number <= pivot.number then this is a new stale head
- else the stale head was already included by some previous finalization.

File: client/service/src/lib.rs (hunk 5 above)
Comment (davxy) [reply client/service/src/lib.rs:155]: The previous behavior was to iterate till the last finalized block and then use that on the `on_block_finalized`.

Because we are now directly receiving the last block I tough that we may get rid of the lookup.

But maybe I don't understood what you are suggesting :-)

Regarding the `on_block_finalized` modification to take the full notification, yes I agree with you that may be a good idea for another issue.

File: client/service/src/client/client.rs (hunk 7 above)
Comment (andresilva) [reply client/service/src/client/client.rs:865]: I think adding a reduced version of this as a comment would be helpful for future readers, particularly the fact that stale heads might not be immediately pruned from `backend.leaves()`.

File: client/service/test/src/client/mod.rs [hunk 8]
```diff
@@ -975,6 +1012,112 @@ fn finalizing_diverged_block_should_trigger_reorg() {
...
+	block_on(client.import(BlockOrigin::Own, d3.clone())).unwrap();
+
+	let d4 = client
+		.new_block_at(&BlockId::Hash(d3.hash()), Default::default(), false)
+		.unwrap()
+		.build()
+		.unwrap()
+		.block;
+	// Postpone import to test behavior of import of finalized block.
+
```
Comment (andresilva) [inline client/service/test/src/client/mod.rs:1108]:
```suggestion

	// Postpone import to test behavior of import of finalized block.
```

**Review by andresilva — APPROVED** [review]
LGTM! Nice work @davxy

File: client/service/src/client/client.rs (hunk 7 above)
Comment (davxy) [reply client/service/src/client/client.rs:874]: Done

File: client/service/src/client/client.rs [hunk 9]
```diff
@@ -831,58 +847,82 @@ where
...
-			for finalized in &enacted[start..] {
-				operation.notify_finalized.push(finalized.hash);
+			let finalized =
+				route_from_finalized.enacted().iter().map(|elem| elem.hash).collect::<Vec<_>>();
+
+			let last_finalized_number = self
+				.backend
+				.blockchain()
+				.number(last_finalized)?
+				.expect("Finalized block expected to be onchain. qed");
```
Comment (bkchr) [inline client/service/src/client/client.rs:857]:
```suggestion
				.expect("Finalized block expected to be onchain; qed");
```

File: client/service/src/client/client.rs [hunk 10]
```diff
@@ -831,58 +847,82 @@ where
...
+				// heads that were in a stale state before this finalization and thus already
+				// included in previous notifications. We want to skip such heads.
+				// Given the "route" from the currently finalized block to the head under
+				// analysis, the condition for it to be added to the new stale heads list is:
+				// `!retracted.is_empty() && !(last_finalized_number > pivot.number)`
+				// 1. "route" has some "retractions".
+				// 2. previously finalized block number is not greater than the "route" pivot:
+				//    - if `last_finalized_number <= pivot.number` then this is a new stale head;
+				//    - else the stale head was already included by some previous finalization.
+				if !(retracted.is_empty() || last_finalized_number > pivot.number) {
```
Comment (bkchr) [inline client/service/src/client/client.rs:874]:
```suggestion
				let route_from_last_finalized =
					sp_blockchain::tree_route(self.backend.blockchain(), last_finalized, head)?;
				
				// If the fork originates from the last finalized block, it is a "fresh" fork.
				// Otherwise it means that we already had reported this fork with an earlier notification. 
				if route_from_last_finalized.retracted().is_empty() {
```

Comment (bkchr) [reply client/service/src/client/client.rs:874]: Regarding this
```
if `last_finalized_number <= pivot.number` then this is a new stale head
```

We should make sure that block import doesn't work for forks that do not originate from last finalized block. We should add some test and the implement this.

**Review by bkchr — COMMENTED** [review]
Looks already very good, but needs some more adjustments.

**Comment by bkchr** [comment]
Ahh and @davxy please don't use force pushes :)

File: client/service/src/client/client.rs (hunk 10 above)
Comment (davxy) [reply client/service/src/client/client.rs:871]: With the suggested modification it will also push heads for forks that originate after the currently finalized block. E.g.:

```
LF: prev finaliized, CF: currenlty finalized

LF----CF---+----->
            \----->  NotStale (this will be pushed)
```
The check `route_from_last_finalized.retracted().is_empty()` has indeed the same function of the numbering check I made, but should be paired with `!route_from_finalized.retracted.is_empty()` to prevent adding branches that are not stale.

I've chosen the numbering technique to prevent computing another route and exploit the one I already have.

---

Regarding this:

> *We should make sure that block import doesn't work for forks that do not originate from last finalized block.*

I will test this condition

File: client/service/src/client/client.rs [hunk 11]
```diff
@@ -683,11 +683,27 @@ where
...
+				};
+				summary.finalized.push(hash);
+				// Add to the stale list all heads that are branching from parent.
+				for head in self.backend.blockchain().leaves()? {
+					if head == parent_hash {
+						continue
+					}
+					let route_from_parent =
+						sp_blockchain::tree_route(self.backend.blockchain(), parent_hash, head)?;
+					if route_from_parent.retracted().is_empty() {
```
[the remark this answers is not in the document]
Comment (davxy) [reply client/service/src/client/client.rs:701]: This is to prevent adding notifications for the heads that we already added in the `apply_finality_with_block_hash`.

At this point (i.e. adding another finalized block on top of the parent and after `apply_finality_with_block_hash`) the only additional stale forks that the `apply_finality_with_block_hash` may have missed are the ones originating from the parent itself without any retraction.

File: client/service/src/client/client.rs [hunk 12]
```diff
@@ -622,25 +622,25 @@ where
...
 
+		// Ensure parent chain is finalized to maintain invariant that finality is called
+		// sequentially.
+		if finalized && parent_exists {
+			self.apply_finality_with_block_hash(
+				operation,
+				parent_hash,
+				None,
+				info.best_hash,
+				make_notifications,
```
[the remark this answers is not in the document]
Comment (davxy) [reply client/service/src/client/client.rs:640]: The at line 694 is the current block hash. The function is invoked using the `parent_hash`. The current hash is not included in this operation.

The current hash is not passed because `apply_finality_with_block_hash` assumes that the blocks to mark as finalized are already stored by the backend and are only marked as final

Snip from `apply_finality...`
```
			operation.op.mark_finalized(BlockId::Hash(finalize_new.hash), None)?;
```

The currently imported block is excluded by from this call because it is directly imported as final.

Snip from `execute_and_import_block`
```
		let leaf_state = if finalized {
			NewBlockState::Final
		} else if is_new_best {
			NewBlockState::Best
		} else {
			NewBlockState::Normal
		};
		...
		operation.op.set_block_data(
			import_headers.post().clone(),
			body,
			indexed_body,
			justifications,
			leaf_state,
		)?;
```

File: client/service/src/client/client.rs (hunk 10 above)
Comment (bkchr) [reply client/service/src/client/client.rs:873]: > ```
> LF: prev finaliized, CF: currenlty finalized
> 
> LF----CF---+----->
>             \----->  NotStale (but this will be pushed)
> ```


Why would the NotStale one pushed here? It is originated from last finalized as well? Or what do I miss here?

Comment (davxy) [reply client/service/src/client/client.rs:873]: Because `route_from_last_finalized.retracted().is_empty()` for the route from `LF` to `NotStale` is true.

File: client/service/src/client/client.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (davxy) [reply client/service/src/client/client.rs:640]: This marks as finalized all the blocks up to the parent. The current block hash is not included here, it will be appended at  line 694.

Current block is not included in the call to `apply_finality_with_block_hash` because this function cannot be used for blocks that are not already stored in the backend.

The currently imported one is treated specially, it is directly imported as Final (row 652)

File: client/service/src/client/client.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (davxy) [reply client/service/src/client/client.rs:701]: All the stale branches and finalized blocks are already included by the `apply_finality_with_block_hash` (above).

At this point, by adding another final block on top of the parent, the only extra stale branches that we may have missed are the ones starting from the parent itself and the condition `route_from_parent.retracted().is_empty()` checks this condition.

What may be missing is to perform this operation only if `parent_exists` is true (otherwise recovering the `tree_route` call will fail).

File: client/service/src/client/client.rs [hunk 13]
```diff
@@ -683,11 +683,27 @@ where
...
+		// We only notify when we are already synced to the tip of the chain
 		// or if this import triggers a re-org
 		if make_notifications || tree_route.is_some() {
 			if finalized {
-				operation.notify_finalized.push(hash);
+				let mut summary = match operation.notify_finalized.take() {
+					Some(summary) => summary,
+					None => FinalizeSummary { finalized: Vec::new(), stale_heads: Vec::new() },
+				};
+				summary.finalized.push(hash);
```
[the remark this answers is not in the document]
Comment (davxy) [reply client/service/src/client/client.rs:694]: This is to add the current block hash. All the others, up to the parent, were already added by `apply_finality_with_block_hash`.

File: client/service/src/client/client.rs (hunk 10 above)
Comment (bkchr) [reply client/service/src/client/client.rs:882]: Yeah. You are right. Not sure why I was not seeing this yesterday :D Ty for the explanation.

Comment (bkchr) [inline client/service/src/client/client.rs:882]:
```suggestion
				if !retracted.is_empty() && last_finalized_number <= pivot.number {
```

File: client/service/src/client/client.rs [hunk 14]
```diff
@@ -831,58 +847,82 @@ where
...
+				let route_from_finalized =
+					sp_blockchain::tree_route(self.backend.blockchain(), block, head)?;
+				let retracted = route_from_finalized.retracted();
+				let pivot = route_from_finalized.common_block();
+				// It is not guaranteed that `backend.blockchain().leaves()` doesn't return
+				// heads that were in a stale state before this finalization and thus already
+				// included in previous notifications. We want to skip such heads.
+				// Given the "route" from the currently finalized block to the head under
+				// analysis, the condition for it to be added to the new stale heads list is:
+				// `!retracted.is_empty() && !(last_finalized_number > pivot.number)`
```
Comment (bkchr) [inline client/service/src/client/client.rs:877]:
```suggestion
				// `!retracted.is_empty() && last_finalized_number <= pivot.number`
```

File: client/service/src/client/client.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (bkchr) [reply client/service/src/client/client.rs:640]: Now I have read the comment above :facepalm:

File: client/service/src/client/client.rs [hunk 15]
```diff
@@ -683,11 +683,27 @@ where
...
 		// or if this import triggers a re-org
 		if make_notifications || tree_route.is_some() {
 			if finalized {
-				operation.notify_finalized.push(hash);
+				let mut summary = match operation.notify_finalized.take() {
+					Some(summary) => summary,
+					None => FinalizeSummary { finalized: Vec::new(), stale_heads: Vec::new() },
+				};
+				summary.finalized.push(hash);
+				// Add to the stale list all heads that are branching from parent.
```
Comment (bkchr) [inline client/service/src/client/client.rs]:
```suggestion
				// Add to the stale list all heads that are branching from parent besides our current `head`.
```

**Review by bkchr — APPROVED** [review]
